Script to generate readonly kubeconfig

dipankardas011 opened the pull request:
## Description
it adds a script for genrating readonly permissions for pods


## Tasks
- Closes #18

cc @nikimanoledaki

maxbischoff reviewed (commented):
Hey, I've seen the topic in the CNCF Slack and thought I might be able to help. Since you already did good work, I just though I can offer these minor suggestions.

dipankardas011 commented on clusters/script-gen-readonly-kubeconfig.sh:
```diff
@@ -0,0 +1,71 @@
+
+cat <<EOF | kubectl apply -f -
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: crole-pod-readyonly
+rules:
+- apiGroups: [""]
+  resources: ["pods"]
+  verbs: ["get", "watch", "list"]
```
the original proposal was cluster wide pod only
here @nikimanoledaki has a say

nikimanoledaki replied:
Thank you @maxbischoff & @dipankardas011 for looking into this! The only strict rule was to exclude Secrets, which the default [`view` ClusterRole](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#user-facing-roles) does as well. I didn't know that there were default out-of-the-box Roles. :)

Limiting to `pod` was so that we can start small if we were to create a Role from scratch. However, if we can benefit from the ease of having a default role out of the box that already excludes Secrets, I don't see any reason against this - @antoniodituri @rossf7 wdyt?

nikimanoledaki replied:
@dipankardas011 did you have a chance to test the script with the default `view` ClusterRole?

dipankardas011 replied:
Okay will check it out
But one question that do we actually need it?
Should the user have read access to all the other resources except secret is it good ?

rossf7 replied:
I'm late to the party but +1 from me on using the default view ClusterRole.

I think secrets are the only resources we need to restrict. Being able to view the rest of the resources including the Flux resources will be useful for trouble shooting.

dipankardas011 replied:
got it 👍🏼

maxbischoff replied:
@rossf7 the default `view` ClusterRole does not grant read access to custom resources, such as Flux resources. It only grants access to core kubernetes resources. You can check this by looking at the ClusterRole using `kubectl get clusterrole view -o yaml`

We'll have to add a second ClusterRole and bind it to grant access to Custom Resources.

rossf7 replied:
@maxbischoff TIL 😄 thank you and I think a second ClusterRole for the Flux custom resources would be useful.

dipankardas011 replied:
added aggregation support

dipankardas011 commented on docs/read-only-kubeconfig.md:
````diff
@@ -0,0 +1,23 @@
+## Steps to get the kubeconfig
+
+```bash
+chmod u+x clusters/script-gen-readonly-kubeconfig.sh 
````
how does that work?

maxbischoff replied:
By executing the command (`chmod u+x clusters/script-gen-readonly-kubeconfig.sh`) and then adding the file to git and commiting it.

dipankardas011 replied:
I thought git doesn't respect file permissions

maxbischoff replied:
It should work https://stackoverflow.com/a/40979016
, I do remember that it didn't work in Windows with directories mounted into WSL back when I used that.

maxbischoff replied:
But this is just a nitpick, I don't believe this should block the Pull Request

dipankardas011 commented on clusters/script-gen-readonly-kubeconfig.sh:
```diff
@@ -0,0 +1,96 @@
+  labels:
+    rbac.authorization.k8s.io/aggregate-to-view: "true"
+rules:
+- apiGroups:
+    - source.toolkit.fluxcd.io/v1beta2
+  resources: 
+    - HelmRepository
+  verbs: 
+    - get
+    - watch
+    - list
+
+- apiGroups: 
+    - helm.toolkit.fluxcd.io/v2beta1
+  resources: 
+    - HelmRelease
+  verbs: 
+    - get
+    - watch
+    - list
+
+- apiGroups: 
+    - kustomize.config.k8s.io/v1beta1
+  resources:
+    - Kustomization
+  verbs: 
+    - get
+    - watch
+    - list
```
Here we can add other resources we want to get access

dipankardas011 replied:
I am not quite sure about these apigroups and resources

dipankardas011 commented on clusters/script-gen-readonly-kubeconfig.sh:
```diff
@@ -0,0 +1,96 @@
+metadata:
+  name: crole-customresources-readyonly
+  labels:
+    rbac.authorization.k8s.io/aggregate-to-view: "true"
```
it aggregates using this label to `view`

https://kubernetes.io/docs/reference/access-authn-authz/rbac/#aggregated-clusterroles

maxbischoff replied:
That makes sense, good find!

maxbischoff commented on clusters/script-gen-readonly-kubeconfig.sh:
```diff
@@ -0,0 +1,96 @@
+    rbac.authorization.k8s.io/aggregate-to-view: "true"
+rules:
+- apiGroups:
+    - source.toolkit.fluxcd.io/v1beta2
```
`apiGroups` need to be provided without the `version` part of the name, so this needs to be:

```suggestion
    - source.toolkit.fluxcd.io
```

maxbischoff commented on clusters/script-gen-readonly-kubeconfig.sh:
```diff
@@ -0,0 +1,96 @@
+    - list
+
+- apiGroups: 
+    - helm.toolkit.fluxcd.io/v2beta1
```
```suggestion
    - helm.toolkit.fluxcd.io
```

maxbischoff commented on clusters/script-gen-readonly-kubeconfig.sh:
```diff
@@ -0,0 +1,96 @@
+    - list
+
+- apiGroups: 
+    - kustomize.config.k8s.io/v1beta1
```
```suggestion
    - kustomize.config.k8s.io
```

maxbischoff reviewed (changes requested):
Looks good to me! Only the apiGroups need to be adjusted, I believe that the version numbers need to be omitted, otherwise the rules won't work.

dipankardas011 commented:
> Looks good to me! Only the apiGroups need to be adjusted, I believe that the version numbers need to be omitted, otherwise the rules won't work.

removed the redundant rules.
when i tried using the rules by default fluxcd came under the view clusterrole scope and it had access

so i removed them and made the `rules: []`

maxbischoff reviewed (approved):
Seems like fluxcd installs a ClusterRole which is aggregated to view then, TIL. Thanks for testing access, looks good to me.

rossf7 commented on clusters/script-gen-readonly-kubeconfig.sh:
```diff
@@ -0,0 +1,73 @@
+CLUSTER_CA=$(kubectl config view --raw -o=go-template='{{range .clusters}}{{if eq .name "'''${CURRENT_CLUSTER}'''"}}"{{with index .cluster "certificate-authority-data" }}{{.}}{{end}}"{{ end }}{{ end }}')
+CLUSTER_SERVER=$(kubectl config view --raw -o=go-template='{{range .clusters}}{{if eq .name "'''${CURRENT_CLUSTER}'''"}}{{ .cluster.server }}{{end}}{{ end }}')
+
+cat << EOF > green-review-cluster-readonly-config
```
```suggestion
cat << EOF > green-reviews-cluster-readonly-config
```

Nitpick but I think we should use plural to match the repo name.

rossf7 commented on docs/read-only-kubeconfig.md:
````diff
@@ -0,0 +1,23 @@
+
+## Test out the kubeconfig
+```bash
+export KUBECONFIG=${PWD}/green-review-cluster-readonly-config
````
```suggestion
export KUBECONFIG=${PWD}/green-reviews-cluster-readonly-config
```

rossf7 reviewed (commented):
@dipankardas011 Just gave this a try and its working great. Nice job!

I just have a doubt on the script location since `./clusters/` is being watched by Flux.

What about moving it to `./scripts/gen-readonly-kubeconfig.sh`?

dipankardas011 commented:
Sure will move it

rossf7 reviewed (approved):
LGTM 🚀
